libsel4vmmplatsupport/src/drivers/pci_helper.c

```diff
@@ -23,12 +23,11 @@ int vmm_pci_mem_device_read(void *cookie, int offset, int size, uint32_t *result
         ZF_LOGE("Offset should not be negative");
         return -1;
     }
-    if (offset + size >= PCI_CAPABILITY_SPACE_OFFSET) {
+    if (offset + size > PCI_CAPABILITY_SPACE_OFFSET) {
         ZF_LOGI("Indexing capability space not yet supported, returning 0");
         *result = 0;
         return 0;
     }
-    *result = 0;
```

Comment:
Is this intentionally removed? I assume it was there to handle when `size < 4` and `*result` isn't 0.

Reply:
Yes it was. And yes, most probably you are correct. And that's a bit problematic. On the other hand the function prototype says the data pointer is uint32_t, but on the another, it also takes the length argument. An ignorant one like me could imagine it's OK to pass a pointer to uint8_t if you pass length = 1, right? But inside this function you get 32-bit write anyway.

I think the best solution would be to change the callback and function prototypes to use void * pointer. That way you could not make any assumptions about the read/write sizes, but you would need to explicitly rely on the length argument. But for now, I just changed the FDT generation code to use uint32_t and also assign the local variables there to zero beforehand, just in case this set-to-zero gets eventually deleted for some reason. So far this works because ARM is configured for little-endian, but that void pointer would make this endian-agnostic. Or so I believe.

TL;DR: added this back, worked around the issues in the FDT code.

Reply:
I agree with your comment. At the moment it's implemented in a way that tries to model the memory region as a 32bit CPU register. To allow 64bit reads/writes this would need to be modified.